src/modules/test/recipes/SecurityBaselineTests.json-footer

```diff
@@ -0,0 +1,4 @@
+  {
+    "Action": "UnloadModule"
+  }
+]
```

Comment:
Concerns with this approach, we need to discuss more:

1. W3e have several other modules that are temporarily disabled from build but which we want to refactor and re-enable later. They may not work with this pattern.

2. In general, test recipes can contain anything. A module test recipe won't be required to have the footer or the header shown here for ASB one. 

3. Test recipes can unload and reload modules any number of times. It's not just one, or two. Thus, what is now split into header, mid and footer may become -if going by this pattern- header, mid1, mid2..., midN, footer.

Last and not least, are the header, footer and mid intended only for the ASB test recipe? If so, why not keep the test recipe one single JSON as it was (and in sync with other such big files, the policy MOF and the MIM, for example) which has the big benefit of keeping the recipe simple?

Reply:
Let me try to answer quickly:

> 1. We have several other modules that are temporarily disabled from build but which we want to refactor and re-enable later. They may or may not work with this pattern.

I don't aim to switch all modules to this pattern, I think it only makes sense where there is a MOF file(s) that map (1 to 1 or many to 1) to a module. Here ASB+SSH MOF are used to test the ASB module. The goal was to get the order and parameters in the test recipe to match the one in a MOF file describing a policy.

> 2. In general, test recipes can contain anything. A module test recipe won't be required to have the footer or the header shown here for ASB one.

Agree

> 3. Test recipes can unload and reload modules any number of times. It's not just one, or two times. Thus, what is now split into header, mid and footer may become (if going by this pattern) header, mid1, mid2..., midN, footer.

I'm not married to the approach - I noticed that the existing test recipe had a unload/reload in the middle between remediate and audit steps, and the reload is necessary to get the SSH settings to be applied.

> Are the header, footer and mid intended only for the ASB test recipe? If so, why not keep the test recipe one single JSON as it was (and in sync with other such big files, the policy MOF and the MIM, for example) which has the big benefit of keeping the recipe simple and easy to understand and maintain in the future?

The generated ASB file is still present: https://github.com/Azure/azure-osconfig/pull/821/files#diff-5a26e381378cc3525d9236fc6e0b756218ceaf04daf0f5f04c380954d994e751 and used. The header/footer/mid parts are just templates used as part of the generation specific to the ASB recipe. Test recipes for other policy modules (CIS) might not need a mid section.

Hope that helps clear things up